### PR TITLE
HDDS-10645. Support x-amz-metadata-directive in CopyObject

### PR DESCRIPTION
**Pull request by ivandika3**
## What changes were proposed in this pull request?

Need to support x-amz-metadata-directive: specifies whether the metadata is copied from the source object or replaced with metadata that's provided in the request. When copying an object, you can preserve all metadata (the default) or specify new metadata. If this header isn’t specified, COPY is the default behavior.

Valid Values: COPY | REPLACE

See: https://docs.aws.amazon.com/AmazonS3/latest/API/API_CopyObject.html#API_CopyObject_RequestSyntax

## What is the link to the Apache JIRA

https://issues.apache.org/jira/browse/HDDS-10645

## How was this patch tested?

Unit test and acceptance test.

Clean CI run: https://github.com/ivandika3/ozone/actions/runs/8589734492

### CODE REVIEW

**Comment by ivandika3** [comment]
@kerneltime @DaveTeng0 @SaketaChalamchala @vtutrinov Could you help take a look when you have time?

File: hadoop-ozone/s3gateway/src/main/java/org/apache/hadoop/ozone/s3/endpoint/ObjectEndpoint.java [hunk 1]
```diff
@@ -1208,12 +1210,30 @@ private CopyObjectResponse copyObject(OzoneVolume volume,
...
+      if (StringUtils.isEmpty(metadataCopyDirective) || metadataCopyDirective.equals(CopyDirective.COPY.name())) {
+        // The custom metadata will be copied from the source key
+        customMetadata = sourceKeyDetails.getMetadata();
+      } else if (metadataCopyDirective.equals(CopyDirective.REPLACE.name())) {
+        // Replace the metadata with the metadata form the request headers
+        customMetadata = getCustomMetadataFromHeaders(headers.getRequestHeaders());
+      } else {
+        OS3Exception ex = newError(INVALID_ARGUMENT, metadataCopyDirective);
+        ex.setErrorMessage("An error occurred (InvalidArgument) " +
+            "when calling the CopyObjet operation: " +
```
Comment (adoroszlai) [inline hadoop-ozone/s3gateway/src/main/java/org/apache/hadoop/ozone/s3/endpoint/ObjectEndpoint.java:1226]:
```suggestion
            "when calling the CopyObject operation: " +
```

File: hadoop-ozone/s3gateway/src/main/java/org/apache/hadoop/ozone/s3/endpoint/ObjectEndpoint.java [hunk 2]
```diff
@@ -1208,12 +1210,30 @@ private CopyObjectResponse copyObject(OzoneVolume volume,
...
+            "The metadata directive specified is invalid. The valid values are COPY or REPLACE.");
```
Comment (adoroszlai) [inline hadoop-ozone/s3gateway/src/main/java/org/apache/hadoop/ozone/s3/endpoint/ObjectEndpoint.java:1227]:
```suggestion
            "The metadata directive specified is invalid. Valid values are COPY or REPLACE.");
```

**Review by adoroszlai — COMMENTED** [review]
Thanks @ivandika3 for the patch.

File: hadoop-ozone/s3gateway/src/main/java/org/apache/hadoop/ozone/s3/endpoint/ObjectEndpoint.java (hunk 1 above)
Comment (ivandika3) [reply hadoop-ozone/s3gateway/src/main/java/org/apache/hadoop/ozone/s3/endpoint/ObjectEndpoint.java]: Thank you for catching it. Updated.

File: hadoop-ozone/s3gateway/src/main/java/org/apache/hadoop/ozone/s3/endpoint/ObjectEndpoint.java (hunk 2 above)
Comment (ivandika3) [reply hadoop-ozone/s3gateway/src/main/java/org/apache/hadoop/ozone/s3/endpoint/ObjectEndpoint.java]: Updated.

**Review by adoroszlai — COMMENTED** [review]
Thanks @ivandika3 for updating the patch, LGTM.  I'll wait some time for another review.

**Review by tanvipenumudy — APPROVED** [review]
Thank you @ivandika3 for the patch, the changes look good.

**Comment by adoroszlai** [comment]
Thanks @ivandika3 for the patch, @tanvipenumudy for the review.

**Comment by ivandika3** [comment]
Thank you for the reviews @adoroszlai @tanvipenumudy

**Pull request merged**